fix: Run HFLocalInvocationLayer.supports even if inference packages are not installed

MichelBartels opened the pull request:
### Related Issues

- #4848 

### Proposed Changes:

HFLocalInvocationLayer.supports would previously fail if the optional inference dependencies are not installed. This PR causes it not to fail and instead to return false.

### How did you test it?
I added a test case in test_hugging_face.py::test_supports.

### Notes for the reviewer

@vblagoje originally discovered the bug and discussed the solution with me as #4848 affects when this problem occurs.

### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.
- I documented my code
- I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

vblagoje commented:
Looks much better @MichelBartels Shall we also replace the get task call in HF inference to use our own get task with timeout?

MichelBartels commented:
Yeah, that makes sense. I have just changed the import and added test, but I'm not sure if it still makes sense for the get_task function to be in `invocation_layer/hugging_face.py`. Do you know of a better place to put the function?

vblagoje commented:
What about utils.py of the invocation layer?

vblagoje commented:
LGTM @MichelBartels if you haven't, would you please run the manual tests with and without inference installed? I'll let @ZanSara give the final blessing to this one.

MichelBartels commented:
While it seems to be working as expected without inference installed, I'm not sure which tests I should run because they are all assuming that torch is available.
I can run test_huggingface_inference.py without a failure, but that is also expected because HuggingFace inference does not require any inference packages.

vblagoje reviewed (changes requested):
@MichelBartels everything seems ok except the LOC where we check for the presence of `api_key` in both hf supports layers. It should be updated to :
```
bool(kwargs.get("api_key", None))
```
in both hugging_face.py and hugging_face_inference.py
Let's add hf and hf inference supports test for setting api_key to "" and None value

vblagoje commented:
@MichelBartels have another look now. The default install, installs transformers and huggingface-hub but not the torch. So I adjusted guards accordingly, fixed the api_key checks in both supports methods, added and adjusted some unit tests. Please have a look yourself.

MichelBartels commented:
Thanks, this is much simpler. I missed that we don't actually need any non-default packages to run the supports methods. I changed the title accordingly.

MichelBartels commented:
I wonder if it still makes sense to check if the inference package is installed before returning True. This could be helpful because it would fail earlier.

vblagoje commented:
It works now. Try it out, please, to double-check. Create a new venv, do the default install from this branch. And then run:

```
from haystack.nodes.prompt import PromptNode

pn = PromptNode(model_name_or_path="google/flan-t5-xxl", api_key="you_hf_api_key")
response = pn("Tell me more Berlin")
print(response)
```

MichelBartels commented on haystack/nodes/prompt/invocation_layer/hugging_face.py:
```diff
@@ -318,14 +309,15 @@ def _extract_torch_dtype(self, **kwargs) -> Optional["torch.dtype"]:
     @classmethod
     def supports(cls, model_name_or_path: str, **kwargs) -> bool:
         task_name: Optional[str] = kwargs.get("task_name", None)
+        if not torch_and_transformers_import.is_successful():
+            return False
```
Should we perhaps remove those two lines again? Considering the following lines work without the inference install, we would get a more meaningful error message without those lines.

vblagoje replied:
Hmm, right. Let me try

vblagoje replied:
You are right. We don't need it. I tried four different options: 
- with and without inference
- hf and hf inference layer

Everything works as expected. In addition, as you suggest, the error is more meaningful in cases when the user intends to use hugging_face.py layer yet forgot to install torch.

MichelBartels commented:
I don't have a HuggingFace Api key so it can't verify that the first example is working, but it fails at the verification stage with a fake key so it seems to correctly identify the layer.
The second example fails as expected.

vblagoje reviewed (approved):
Good to go now.
